[AArch64] merge index address with large offset into base address

vfdff opened the pull request:
A case for this transformation, https://gcc.godbolt.org/z/nhYcWq1WE
Fold
```
  mov     w8, #56952
  movk    w8, #15, lsl #16
  ldrb    w0, [x0, x8]
```
into
```
  add     x0, x0, 1036288
  ldrb    w0, [x0, 3704]
```

Only LDRBBroX is supported for the first time.
Note: The AArch64DAGToDAGISel is difficult to check the loop invariant, which may bring in regression.
Fix https://github.com/llvm/llvm-project/issues/71917

vfdff committed:
[AArch64] merge index address with large offset into base address

A case for this transformation, https://gcc.godbolt.org/z/nhYcWq1WE
Fold
  mov     w8, #56952
  movk    w8, #15, lsl #16
  ldrb    w0, [x0, x8]
into
  add     x0, x0, 1036288
  ldrb    w0, [x0, 3704]

Only LDRBBroX is supported for the first time.
Fix https://github.com/llvm/llvm-project/issues/71917

david-arm commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -4089,6 +4089,11 @@ AArch64InstrInfo::getLdStOffsetOp(const MachineInstr &MI) {
 
+const MachineOperand &
+AArch64InstrInfo::getLdStAmountOp(const MachineInstr &MI) {
+  return MI.getOperand(4);
```
How do you know this is always going to be operand 4 without inspecting the opcode? For example, see the functions above.

vfdff replied:
yes, it need extend with a `switch case ` when we add more supported instrunction. 
The first version I only supported the `LDRBBroX `,  and it is **operand 4**.
I only add one instrunction in the first version to make sure if it is reasonable for the optimization direction itself.
If it is adopted, I will add other memory access instructions in the subsequent patch.

david-arm replied:
OK, but my point is that if your patch lands another engineer may start using this function believing that it is complete. However, it might do the wrong thing if called for a different opcode. I think you need to explicitly add an error if the opcode is not what you expect so that it fails sensibly.

vfdff replied:
Thanks for your idea, fixed with commit  01029d9

david-arm commented on llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:
```diff
@@ -481,6 +505,16 @@ static unsigned getPreIndexedOpcode(unsigned Opc) {
+  // TODO: Add more index address loads/stores.
+  switch (Opc) {
+  default:
+    llvm_unreachable("Opcode has no base address equivalent!");
```
At first glance this doesn't look right to me. Are there any checks to actually prevent us getting here with other opcodes? If not, then I think you may need to fill in the remaining opcodes in the switch statement, similar to `getPostIndexedOpcode`.

vfdff replied:
yes, similar to above getLdStAmountOp，I only support `LDRBBroX` in the first version.

david-arm replied:
I guess the point I'm trying to make here is that usually we add a llvm_unreachable because we actually expect the code to be unreachable. However, I can't see anything that prevents us hitting the default case. For example, I'd expect code to be written in this way:

```
void foo(MI) {
  if (opcode == my_expected_opcode)
    foo2(MI);
}

void foo2() {
  switch (MI.getOpcode()) {
  default:
    llvm_unreachable("Unexpected opcode");
  case my_expected_opcode:
    do_something();
    break;
  }
}
```

My question was - do you have anything to explicitly prevent us from calling the function with an unexpected opcode?

vfdff replied:
yes, the function `isMergeableIndexLdSt` called by `AArch64LoadStoreOpt::optimizeBlock` will check only the expected opcode will try this fold. (Now only support one opcode `LDRBBroX` )

vfdff committed:
[AArch64] Fold more load.x into load.i with large offset

The list of load.x is refer to canFoldIntoAddrMode on D152828.
Also support LDRSroX missed in canFoldIntoAddrMode

vfdff commented:
support more load.x instruction, the list is refered to `canFoldIntoAddrMode`

vfdff commented:
ping ?

davemgreen commented on llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:
```diff
@@ -2220,6 +2395,60 @@ MachineBasicBlock::iterator AArch64LoadStoreOpt::findMatchingUpdateInsnBackward(
+
+  // Make sure the IndexReg is killed and the shift amount is zero.
+  // TODO: Relex this restriction to extend, simplify processing now.
+  if (!AArch64InstrInfo::getLdStOffsetOp(MemMI).isKill() ||
```
Is this essentially a one-use check on the MOV reg?

vfdff replied:
yes, because we'll delete these old instructions when we math the folding in the end of function **mergeConstOffsetInsn**.

davemgreen reviewed (commented):
Yeah thanks, I think this is looking better now.

Can you add some tests for different types and edge conditions?

vfdff commented on llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:
```diff
@@ -171,6 +178,13 @@ struct AArch64LoadStoreOpt : public MachineFunctionPass {
 
+  // Scan the instruction list to find a register assigned with a const
+  // value that can be combined with the current instruction (a load or store)
+  // using base addressing with writeback. Scan forwards.
```
yes, you are right, I'll update

vfdff commented:
> Can you add some tests for different types and edge conditions?
Done, thanks

davemgreen commented on llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:
```diff
@@ -722,6 +781,40 @@ static bool isMergeableLdStUpdate(MachineInstr &MI) {
+    Scale = 2;
+    return true;
+  case AArch64::LDRWroX:
+  case AArch64::LDRSWroX:
```
Should there be LDRSroX too?

vfdff replied:
Apply your comment, thanks

davemgreen reviewed (commented):
Thanks. I noticed LDRSroX is missing, but so long as the perf looks OK with that added this looks OK.

vfdff committed:
[AArch64] Precommit tests for PR75343, NFC

vfdff commented on llvm/test/CodeGen/AArch64/arm64-addrmode.ll:
```diff
@@ -295,3 +459,27 @@ define i32 @LdOffset_i16_odd_offset(ptr nocapture noundef readonly %a)  {
   ret i32 %conv
 }
 
+; Already encoded with a signle mov MOVNWi
```
Done

vfdff commented:
> Thanks. I noticed LDRSroX is missing, but so long as the perf looks OK with that added this looks OK.
Add the missing LDRSroX and releated tests, thanks.

davemgreen reviewed (approved):
Thanks. Every time we write patches like this they end up going wrong in some way. Please make sure you've run some decent testing and keep an eye if this might have gone wrong. From what I can tell it LGTM though. Cheers

asmok-g commented:
heads-up : ) this is causing some tests crashing internally in google. Trying to come up with a reproducer.

vfdff commented:
> heads-up : this is causing some tests crashing internally in google. Trying to come up with a reproducer.
Thank you for your feedback. I will analyze this problem after you provide a method to reproduce the problem ASAP.

davemgreen commented:
It looks like this is causing #79756 too. I hadn't realised that it was committed in separate commits and then only partially reverted. I'll remove the LDRBBroX part too to make sure we don't see problems on the llvm-19 branch.

@vfdff there is a test case in #79756 that looks useful for figuring out at least one of the problems going on.

vfdff commented:
@davemgreen  Thanks very much, I'll take a look at it.
